pkg/ast/ast_operation_definition.go

```diff
@@ -67,3 +60,23 @@ func (d *Document) AddVariableDefinitionToOperationDefinition(operationDefinitio
 	d.OperationDefinitions[operationDefinitionRef].VariableDefinitions.Refs =
 		append(d.OperationDefinitions[operationDefinitionRef].VariableDefinitions.Refs, ref)
 }
+
+const (
+	alphabet = `abcdefghijklmnopqrstuvwxyz`
+)
+
+func (d *Document) GenerateUnusedVariableDefinitionName(operationDefinition int) []byte {
```

Comment:
I do not quite get the purpose of using alphabet. Your goal is to have some unique ID, but essentially why not just to use the array and store index somewhere. You do not think that int64 will be enough?

Reply:
According to the GraphQL specification a variable definition needs to start with a letter. (http://spec.graphql.org/draft/#Name) As I want to be able to create a name as short as possible but also unique I thought I just pick something from the alphabet until it is unique. Got a better idea?

Reply:
Ah, did not realize that it is part of the spec.
Well the more calls to this function, in context of the given document, the more cycles you will have to do, since complexity will be growing linearly. 
You can simplify it to smth like:


```go
const (
	alphabet = `abcdefghijklmnopqrstuvwxyz0123456789`
)

func (d *Document) GenerateUnusedVariableDefinitionName(operationDefinition int) []byte {
	out := make([]byte, 0, 2)
	for i := 0; true; i++ {
		out = append(out, alphabet[rand.Intn(len(alphabet))])
		_, exists := d.VariableDefinitionByNameAndOperation(operationDefinition, out)

		if !exists {
			return out[:i]
		}
	}
	return nil
}
```

So essentially out is pre-allocated outside of the cycle with some default cap, and we take a random letter to reduce the conflict chance. Also, you can increase the alphabet with numbers for the same reason. 

Benchmark do not show any improvements, but I think it is mostly to its synthetic nature. 
If you will use smth like Shopify schema for testing I believe benefit will be move visible.e

Reply:
From a performance point of view I understand your point, makes total sense.
One question regarding your proposal. As the name is not deterministic how would you make this test pass with your implementation?
https://github.com/TykTechnologies/graphql-go-tools/blob/0ab7577b6197016be706d3306c5b7f4ef07af7fd/pkg/astnormalization/astnormalization_test.go#L108

Reply:
Can you mock the random function in tests with a seed that is a known stream of bytes?

Reply:
The idea is really smart! We could use an interface for the generator. Then we can test that the actual implementation generates truly random names separately while we're using a mock during e2e testing.

Reply:
@joshblakeley You're stepping in implementing this?

Reply:
I'd love to but it is heavily dependent on the expected turnaround time :)

Reply:
Actually @jensneuse Leons code will produce the exact same sequence of letters every time it will run so your test should be fine.
In production if we really want a pseudorandom sequence we need to seed the rand function first.

Reply:
Make sense! In tests you can even set fixed rand seed.

Reply:
I was thinking about this a bit. While it makes sense to have a faster algorithm for finding a unique name in this case it is counter intuitive to actually use it. This function only runs at "planning time" and the result gets cached. It's more importantly to have an upstream query as short as possible. Because of that it makes more sense to linearly search for the smallest possible name, single character if possible. Execution performance will not be affected. Just the payload to the upstream will be smaller.